<a href="https://colab.research.google.com/github/JorgeEncinas/pytorch_ztm/blob/main/00_pt_fundamentals_redone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 00. PyTorch Fundamentals

In [1]:
print("Hello I'm excited to learn Pytorch!")

Hello I'm excited to learn Pytorch!


In [2]:
!nvidia-smi

Sat May 31 18:38:10 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   48C    P8             11W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [1]:
import torch
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
print(torch.__version__)

2.6.0+cu124


cu refers to the CUDA version.
Pytorch build uses cuda 12.4 probably.

### Cuda Toolkit
CUDA is what enables us to run on NVIDIA GPUs

## Introduction to Tensors
A Tensor is a way to represent data, especially multidimensional data in a numeric format.

There are different kinds of tensors.

In [3]:
# Scalar
scalar = torch.tensor(7)
scalar

tensor(7)

In [4]:
scalar.ndim #No dimensions, just a single number, but still a tensor.

0

In [5]:
scalar.item() #Returns our tensor back as a Python integer.

7

In [6]:
vector = torch.tensor([7, 7]) #Vector

In [7]:
vector.ndim

1

In [8]:
scalar.shape

torch.Size([])

In [9]:
vector.shape

torch.Size([2])

In [10]:
some_other_vector = torch.tensor([1])

In [11]:
some_other_vector.shape

torch.Size([1])

In [12]:
## MATRIX
MATRIX = torch.tensor([[7, 8],
                       [9, 10]])

In [13]:
MATRIX.ndim

2

In [14]:
MATRIX.shape

torch.Size([2, 2])

In [17]:
print(MATRIX[1])
print(MATRIX[0])

tensor([ 9, 10])
tensor([7, 8])


In [21]:
## TENSOR
TENSOR = torch.tensor([[[1, 2, 3],
                       [4, 5, 6],
                       [7, 8, 9]]])
TENSOR

tensor([[[1, 2, 3],
         [4, 5, 6],
         [7, 8, 9]]])

In [22]:
TENSOR.ndim

3

In [23]:
TENSOR.shape

torch.Size([1, 3, 3])

In [24]:
TENSOR[0]

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])

In [25]:
TENSOR2 = torch.tensor([[[2, 3],
                         [4, 5],
                         [6, 7]],
                        [[10, 11],
                         [12, 13],
                         [14, 15]]])
TENSOR2

tensor([[[ 2,  3],
         [ 4,  5],
         [ 6,  7]],

        [[10, 11],
         [12, 13],
         [14, 15]]])

In [26]:
TENSOR2.shape

torch.Size([2, 3, 2])

In [27]:
TENSOR2.ndim

3

### RANDOM TENSORS

Why random tensors?

Important in PyTorch because many NNs start with randomized weights that are later adjusted and fine-tuned (I think it helps with some issues that would happen if the weights were initialized to 0)

Start with random numbers -> look at data -> update random numbers -> look at data -> update again ...

In [29]:
# Create a random tensor of size (3, 4)
random_tensor = torch.rand((3, 4))
random_tensor

tensor([[0.0264, 0.0967, 0.8016, 0.1799],
        [0.1474, 0.2774, 0.8805, 0.8397],
        [0.5833, 0.0259, 0.2448, 0.2025]])

In [30]:
random_tensor.ndim

2

In [34]:
other_rnd = torch.rand((10, 10, 10))
print(other_rnd.ndim)
print(other_rnd.shape)

3
torch.Size([10, 10, 10])


In [ ]:
# Create a random tensor with a similar shape to an image tensor
## A lot of the time, turning images into tensors...


In [35]:
random_image_size_tensor = torch.rand(size=(224, 224, 3)) #height, width, color channel
random_image_size_tensor.shape, random_image_size_tensor.ndim

(torch.Size([224, 224, 3]), 3)

In [37]:
other_similar = torch.rand(size=random_image_size_tensor.shape)
other_similar.shape, other_similar.ndim

(torch.Size([224, 224, 3]), 3)

In [ ]:
# You can also have color channels at the beginning, instead of at the end, like we did here.

In [38]:
zeros = torch.zeros((3, 4))
zeros

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [39]:
ones = torch.ones((3, 4))
ones

tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])

In [40]:
zeros*ones

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])

In [42]:
print(ones.dtype)
print(zeros.dtype)
print(random_image_size_tensor.dtype)

torch.float32
torch.float32
torch.float32


### Creating a range of tensors and tensors-like

In [43]:
torch.range(0,10)

<ipython-input-43-dfdec9b83f7d>:1: UserWarning: torch.range is deprecated and will be removed in a future release because its behavior is inconsistent with Python's range builtin. Instead, use torch.arange, which produces values in [start, end).
  torch.range(0,10)


tensor([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

In [44]:
torch.arange(0,10)

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [45]:
one_to_ten = torch.arange(start=1, end=11, step=1)
one_to_ten

tensor([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [47]:
# Replicate a particular shape of a tensor, but don't want to specify that shape.
ten_zeros = torch.zeros_like(input=one_to_ten)
ten_zeros

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [49]:
print(one_to_ten.shape)
print(ten_zeros.shape)
print(one_to_ten.ndim)
print(ten_zeros.ndim)

torch.Size([10])
torch.Size([10])
1
1


### Tensor Datatypes

In [2]:
#Float 32 Tensor
float_32_t = torch.tensor([3.0, 6.0, 9.0], dtype=None)
float_32_t, float_32_t.dtype

(tensor([3., 6., 9.]), torch.float32)

In [3]:
float_16_t = torch.tensor([3.0, 6.0, 9.0], dtype=torch.float16)
float_16_t, float_16_t.dtype

(tensor([3., 6., 9.], dtype=torch.float16), torch.float16)

In [4]:
some_tensor = torch.tensor(
  [1.0, 2.0, 3.0],
  dtype  = None, # Numerical precision
  device = None, # By default it's "cpu". What device your tensor is on
  requires_grad=False #Whether or not to track the gradients
)

### Tensor datatypes

PRECISION
32 - Single precision
16 - Half precision

How much precision a data is stored with in memory.


## Three big errors you'll find
1. Tensors not the right datatype
2. Tensors not the right shape
3. Tensors not on the right device

DATATYPE / SHAPE / DEVICE

### DEVICE
By default it's "cpu", but it can also be set to "cuda".
This device, if you try to do operations btw tensors living on different devices,

### GRAD
If you want PyTorch to keep track the gradients of a tensor when it goes through certain operations.

In [5]:
float_16_tensor = float_32_t.type(torch.float16)
float_16_tensor

tensor([3., 6., 9.], dtype=torch.float16)

In [6]:
float_16_tensor * float_32_t

tensor([ 9., 36., 81.])

In [7]:
int_32_t = torch.tensor([3, 2, 1], dtype=torch.int32)

In [8]:
float_32_t * int_32_t

tensor([ 9., 12.,  9.])

### Getting information from Tensors

Datatype, shape, or device

In [11]:
print(int_32_t.dtype) #Tells you the type
print(int_32_t.shape) #Tells you the shape
print(int_32_t.device) #Tells you the device the tensor's on.

torch.int32
torch.Size([3])
cpu


### Manipulating Tensors (tensor operations)

Tensor operations include:
- Addition
- Subtraction
- Multiplication (element-wise)
- Division
- Matrix Multiplication

A NN will use tensors and apply these operations to train the model and do inferences.

In [12]:
tensor = torch.tensor([1, 2, 3])
tensor + 10

tensor([11, 12, 13])

In [13]:
tensor * 10

tensor([10, 20, 30])

In [15]:
tensor - 10

tensor([-9, -8, -7])

In [16]:
tensor / 10

tensor([0.1000, 0.2000, 0.3000])

In [17]:
torch.mul(tensor, 10)

tensor([10, 20, 30])

In [20]:
torch.mul(tensor, torch.tensor([2, 3, 4]))

tensor([ 2,  6, 12])

### Matrix Multiplication

There are two main ways of performing multiplication in NNs/Deep Learning.

1. Element-wise multiplication
2. Matrix Multiplication. (Dot Product)

In [21]:
tensor * tensor #Element-wise Multiplication

tensor([1, 4, 9])

In [22]:
tensor @ tensor #Matrix Multiplication
#You can also use torch.matmul() or torch.mm()

tensor(14)

In [23]:
mt = torch.tensor([[1, 2, 3],
                   [4, 5, 6]])
mt2 = torch.tensor([[7, 8],
                    [9, 10],
                    [11,12]])

In [24]:
mt @ mt2 # (2, 3) * (3, 2) = (2, 2)

tensor([[ 58,  64],
        [139, 154]])

### Matrix Multiplication
You can use either @ or torch.matmul()

1. **Inner dimensions** must match
2. The resulting vector will have the shape of the **Outer Dimensions**

One of the most common errors you'll encounter is that the inner dimensions don't match.

In [28]:
torch.rand((3,2)).shape

torch.Size([3, 2])

## Dealing with Shape Errors

In [29]:
tensor_A = torch.tensor([[1, 2],
                         [3, 4],
                         [5, 6]])
tensor_B = torch.tensor([[7, 10],
                         [8, 11],
                         [9, 12]])

In [30]:
torch.mm(tensor_A, tensor_B)

RuntimeError: mat1 and mat2 shapes cannot be multiplied (3x2 and 3x2)

You can fix it by manipulating the shape of one of our tensors, using **transpose**

tensor_A.T

A transpose switches the axes or dimensions of a given tensor

In [31]:
tensor_A.T @ tensor_B

tensor([[ 76, 103],
        [100, 136]])

In [32]:
tensor_A @ tensor_B.T #These are different, see?

tensor([[ 27,  30,  33],
        [ 61,  68,  75],
        [ 95, 106, 117]])

In [33]:
tensor_B.T @ tensor_A #Even this one is still a bit different, even if it's the same matrix, transposed.

tensor([[ 76, 100],
        [103, 136]])

In [36]:
tensor_BAT = tensor_B @ tensor_A.T
print(tensor_BAT)
print(tensor_BAT.T)
print(tensor_A @ tensor_B.T)

tensor([[ 27,  61,  95],
        [ 30,  68, 106],
        [ 33,  75, 117]])
tensor([[ 27,  30,  33],
        [ 61,  68,  75],
        [ 95, 106, 117]])
tensor([[ 27,  30,  33],
        [ 61,  68,  75],
        [ 95, 106, 117]])


### Finding Min, Max, Mean, Sum, etc. (Tensor Aggregation)


In [37]:
x = torch.arange(0, 100, 10)
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

In [38]:
torch.max(x), x.max()

(tensor(90), tensor(90))

In [39]:
torch.min(x), x.min()

(tensor(0), tensor(0))

In [42]:
torch.mean(x, dtype=torch.float32)

tensor(45.)

In [44]:
torch.mean(x.type(torch.float32))

tensor(45.)

In [45]:
torch.mean(x) #seems like torch.mean() needs some specification of the type in some way or another.

RuntimeError: mean(): could not infer output dtype. Input dtype must be either a floating point or complex dtype. Got: Long

In [46]:
torch.sum(x), x.sum()

(tensor(450), tensor(450))

In [48]:
torch.argmin(x), torch.argmax(x) #This one yields the index.

(tensor(0), tensor(9))

### Reshaping, stacking, squeezing and unsqueezing

* Reshaping - reshapes an input tensor to a defined shape
* View - return a view of an input tensor of certain shape but keep the same memory as the original tensor.


Reshaping and view are similar, but view keeps the same memory: same tensor but from a different perspective

* Stacking - combine multiple tensors on top of each other (vstack) or side by side (hstack)
* Squeeze - Removes all '1' dimensions from a tensor
* Unsqueeze - Adds a '1' dimension to a target tensor
* Permute - Return a view of the input with dimensions permuted (swapped) in a certain way

In [4]:
import torch
x = torch.arange(start=1., end=10., step=1)
x

tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [5]:
# Add an extra dimension
# The shape must be compatible!
x_reshaped = x.reshape(1, 7)

RuntimeError: shape '[1, 7]' is invalid for input of size 9

In [7]:
x_reshaped = x.reshape(1, 9)
x_reshaped

tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.]])

In [8]:
x_reshaped.reshape(2,9)

RuntimeError: shape '[2, 9]' is invalid for input of size 9

In [10]:
x_reshaped.reshape(9,1)

tensor([[1.],
        [2.],
        [3.],
        [4.],
        [5.],
        [6.],
        [7.],
        [8.],
        [9.]])

In [12]:
y = torch.arange(start=1., end=11., step=1)
y

tensor([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

In [13]:
y.reshape(5, 2)

tensor([[ 1.,  2.],
        [ 3.,  4.],
        [ 5.,  6.],
        [ 7.,  8.],
        [ 9., 10.]])

In [18]:
z = y.view(10,1)
z, z.shape

(tensor([[ 1.],
         [ 2.],
         [ 3.],
         [ 4.],
         [ 5.],
         [ 6.],
         [ 7.],
         [ 8.],
         [ 9.],
         [10.]]),
 torch.Size([10, 1]))

In [19]:
y, z

(tensor([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.]),
 tensor([[ 1.],
         [ 2.],
         [ 3.],
         [ 4.],
         [ 5.],
         [ 6.],
         [ 7.],
         [ 8.],
         [ 9.],
         [10.]]))

In [22]:
z[:, 0] = 5. #Modifying the view changes the 'y' tensor! And look, the first selector is for row, so you selected ALL rows.
z, y

(tensor([[5.],
         [5.],
         [5.],
         [5.],
         [5.],
         [5.],
         [5.],
         [5.],
         [5.],
         [5.]]),
 tensor([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.]))

In [25]:
# Stacking Tensors
x_stacked = torch.stack([x, x, x, x]) #, dim=0
x_stacked

tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.],
        [1., 2., 3., 4., 5., 6., 7., 8., 9.],
        [1., 2., 3., 4., 5., 6., 7., 8., 9.],
        [1., 2., 3., 4., 5., 6., 7., 8., 9.]])

In [28]:
x_stacked = torch.stack([x, x, x, x], dim=1)
x_stacked, x_stacked.T

(tensor([[1., 1., 1., 1.],
         [2., 2., 2., 2.],
         [3., 3., 3., 3.],
         [4., 4., 4., 4.],
         [5., 5., 5., 5.],
         [6., 6., 6., 6.],
         [7., 7., 7., 7.],
         [8., 8., 8., 8.],
         [9., 9., 9., 9.]]),
 tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.],
         [1., 2., 3., 4., 5., 6., 7., 8., 9.],
         [1., 2., 3., 4., 5., 6., 7., 8., 9.],
         [1., 2., 3., 4., 5., 6., 7., 8., 9.]]))

In [30]:
x2_stacked = torch.vstack([x, x, x, x])
x2_stacked

tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.],
        [1., 2., 3., 4., 5., 6., 7., 8., 9.],
        [1., 2., 3., 4., 5., 6., 7., 8., 9.],
        [1., 2., 3., 4., 5., 6., 7., 8., 9.]])

In [31]:
x3_stacked = torch.hstack([x, x, x, x])
x3_stacked

tensor([1., 2., 3., 4., 5., 6., 7., 8., 9., 1., 2., 3., 4., 5., 6., 7., 8., 9.,
        1., 2., 3., 4., 5., 6., 7., 8., 9., 1., 2., 3., 4., 5., 6., 7., 8., 9.])

### Squeeze, Unsqueeze, Permute

In [32]:
x = torch.zeros(2, 1, 2, 2, 1)
x

tensor([[[[[0.],
           [0.]],

          [[0.],
           [0.]]]],



        [[[[0.],
           [0.]],

          [[0.],
           [0.]]]]])

In [34]:
x.squeeze()

tensor([[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]])

In [36]:
x_reshaped, x_reshaped.shape

(tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.]]), torch.Size([1, 9]))

In [38]:
xr_squeeze = x_reshaped.squeeze()
xr_squeeze, xr_squeeze.shape

(tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.]), torch.Size([9]))

In [42]:
xr_unsq = xr_squeeze.unsqueeze(dim=0)
print(xr_squeeze, xr_squeeze.shape)
xr_unsq, xr_unsq.shape

tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.]) torch.Size([9])


(tensor([[1., 2., 3., 4., 5., 6., 7., 8., 9.]]), torch.Size([1, 9]))

In [43]:
xr_unsq = xr_squeeze.unsqueeze(dim=1) # Look at how this will shift how your data looks!
print(xr_squeeze, xr_squeeze.shape)
xr_unsq, xr_unsq.shape

tensor([1., 2., 3., 4., 5., 6., 7., 8., 9.]) torch.Size([9])


(tensor([[1.],
         [2.],
         [3.],
         [4.],
         [5.],
         [6.],
         [7.],
         [8.],
         [9.]]),
 torch.Size([9, 1]))

### Permute
Rearranges the DIMENSIONS in a specified order.

In [48]:
x = torch.arange(1., 11.)
x = x.unsqueeze(dim=0)
x

tensor([[ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.]])

In [56]:
x_original = torch.rand(size=(128, 224, 3)) # [height, width, channels]
x_original

tensor([[[0.6629, 0.4179, 0.5810],
         [0.8904, 0.6646, 0.1326],
         [0.6914, 0.7646, 0.1470],
         ...,
         [0.7477, 0.5969, 0.7222],
         [0.8245, 0.3343, 0.9319],
         [0.0841, 0.4039, 0.8731]],

        [[0.6852, 0.7271, 0.7344],
         [0.1407, 0.5572, 0.9704],
         [0.8682, 0.7077, 0.9903],
         ...,
         [0.9752, 0.9410, 0.5364],
         [0.8727, 0.7893, 0.9617],
         [0.4595, 0.9616, 0.1652]],

        [[0.7482, 0.9990, 0.1401],
         [0.7189, 0.2356, 0.6555],
         [0.3365, 0.6069, 0.5145],
         ...,
         [0.5131, 0.4334, 0.5033],
         [0.7386, 0.7487, 0.5532],
         [0.5816, 0.4106, 0.2305]],

        ...,

        [[0.3012, 0.0411, 0.1495],
         [0.2373, 0.9324, 0.9643],
         [0.0202, 0.0514, 0.0298],
         ...,
         [0.2737, 0.5545, 0.3681],
         [0.1800, 0.4157, 0.6337],
         [0.2673, 0.9119, 0.6820]],

        [[0.1112, 0.6336, 0.4482],
         [0.4358, 0.3363, 0.3521],
         [0.

In [57]:
# Now let's permute that!
x_permuted = x_original.permute(2, 1, 0) #Note you need to know the dimensions your tensor has, so that you specify them here!
x_original.shape, x_permuted.shape

(torch.Size([128, 224, 3]), torch.Size([3, 224, 128]))

In [58]:
x_permuted2 = x_original.permute(2, 0, 1)
x_original.shape, x_permuted2.shape

(torch.Size([128, 224, 3]), torch.Size([3, 128, 224]))

In [60]:
x = torch.tensor([ [[2, 3],
                    [4, 5],
                    [6, 7]],
                   [[10, 11],
                    [12, 13],
                    [14, 15]]])
x.shape, x.ndim

(torch.Size([2, 3, 2]), 3)

In [61]:
#Okay so here we got two matrices of shape (3, 2)
xper = x.permute(0, 2, 1)
xper.shape, xper

(torch.Size([2, 2, 3]),
 tensor([[[ 2,  4,  6],
          [ 3,  5,  7]],
 
         [[10, 12, 14],
          [11, 13, 15]]]))

In [62]:
inner_matrix = torch.tensor([[2, 3],
                             [4, 5],
                             [6, 7]])
inner_matrix, inner_matrix.T

(tensor([[2, 3],
         [4, 5],
         [6, 7]]),
 tensor([[2, 4, 6],
         [3, 5, 7]]))

In [63]:
xper2 = x.permute(2, 1, 0)
xper2.shape, xper2 #This is harder to understand, isn't it?

(torch.Size([2, 3, 2]),
 tensor([[[ 2, 10],
          [ 4, 12],
          [ 6, 14]],
 
         [[ 3, 11],
          [ 5, 13],
          [ 7, 15]]]))

We can see that what before were two rows of a matrix, now are in separate matrices, and have become columns.
We have two matrices, of three rows each, and two columns each. It seems like the contents of the matrices have been mixed in half.

In [64]:
x_original[0,0,0] = 728218
x_original[0,0,0], x_permuted[0,0,0] #IMPORTANT! This means your permute is a VIEW

(tensor(728218.), tensor(728218.))

### INDEXING (selecting data from tensors)
Similar to numpy

In [66]:
import torch
x = torch.arange(1, 10).reshape(1, 3, 3)
x

tensor([[[1, 2, 3],
         [4, 5, 6],
         [7, 8, 9]]])

In [67]:
x[0]

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])

In [68]:
x[0][0]

tensor([1, 2, 3])

In [69]:
x[0][0][0]

tensor(1)

In [70]:
x.shape, x.ndim

(torch.Size([1, 3, 3]), 3)

In [75]:
# ':' character allows you to select all in a specific dimension.
x[:, 0, :], x[:, 0]

(tensor([[1, 2, 3]]), tensor([[1, 2, 3]]))

In [72]:
x[:, :, 0]

tensor([[1, 4, 7]])

In [79]:
x[0][1][1], x[:, 1, 1].squeeze()

(tensor(5), tensor(5))

In [81]:
# Challenge: return 9, then return 3, 6, 9
x[0][2][2], x[:,2,2]

(tensor(9), tensor([9]))

In [84]:
x[0,:,2]

tensor([3, 6, 9])

### Numpy
Popular scientific python numerical computing library

- PyTorch has compatibility with it. Use torch.from_numpy(ndarray)
- You can do the opposite, torch.Tensor.numpy()

In [85]:
import torch
import numpy as np
array = np.arange(1.0, 8.0)
tensor = torch.from_numpy(array)
array, tensor

(array([1., 2., 3., 4., 5., 6., 7.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

Note that pytorch's default data type is float32, while for numpy it's float64. If you're going from one form to another, take care on what data type you're dealing with.

So how do we change the data type here?

In [87]:
tensor2 = torch.from_numpy(array).type(torch.float32)
tensor2.dtype, tensor2

(torch.float32, tensor([1., 2., 3., 4., 5., 6., 7.]))

In [88]:
array = array + 1
array, tensor #Note that tensor is NOT a view, since it did not change even if array did.

(array([2., 3., 4., 5., 6., 7., 8.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

In [89]:
tensor = torch.ones(7)
numpy_tensor = tensor.numpy()
tensor, numpy_tensor

(tensor([1., 1., 1., 1., 1., 1., 1.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

In [90]:
tensor = tensor + 1
tensor, numpy_tensor #See here that vice versa it isn't a view either!

(tensor([2., 2., 2., 2., 2., 2., 2.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

### Reproducibility (trying to make random out of random)

In short how a neural network learns:

Start with RND #s -> Tensor Operations -> Update Random #s to make them aligned with the data -> repeat

Reproducibility is setting a seed to make it pseudo-random so that other users/readers can achieve the expected outcome that you got in your experiment or code.

## Random Seed
"Flavors" the randomness

In [96]:
import torch

RANDOM_SEED = 42
torch.manual_seed(seed=RANDOM_SEED)

In [97]:
random_tensor_A = torch.rand(2, 2)
random_tensor_B = torch.rand(2, 2)
random_tensor_A, random_tensor_B

(tensor([[0.8823, 0.9150],
         [0.3829, 0.9593]]),
 tensor([[0.3904, 0.6009],
         [0.2566, 0.7936]]))

If you call the two previous cells in order, again and again, you'll get the same tensors over and over! That means our code using them would be reproducible. If you only run the second cell, however, you will keep getting different tensors, so be careful!

### Different Devices: Running Tensors and PyTorch objects in GPUs (and making faster computations)

CUDA + NVIDIA hardware = PyTorch behind the scenes are optimized to work in GPUs

In [98]:
### 1. Getting a GPU
# You can set it no Google Colab for free. But there is a paid plan for 10 dollars a month (O_o)
### 2. Use your own GPU
# Needs investment ($$$) and setup.
# Tim Dettmers - Which GPU for Deep Learning
### 3. Use Cloud Computing - GCP, AWS, Azure. Rent computers on the cloud and access them.
# Careful with getting a big bill

#For 2, 3
# PyTorch documentation for starting locally

### Changing runtime
We do Runtime > Change runtime type > T4 GPU

In [1]:
!nvidia-smi

Wed Jun  4 01:21:47 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   66C    P8             11W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [3]:
# Check for GPU access with PyTorch
import torch
torch.cuda.is_available()

True

So, good! Google Colab gives us access to a GPU!

In [4]:
# Setup device agnostic code.
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

Look at PyTorch's documentation on Device Agnostic Code.
But in essence, it's best practice to do this every time! Try to use gpu, otherwise default to cpu


### Putting Tensors and Models on the GPU

In [7]:
# Count number of devices
torch.cuda.device_count()

1

In [11]:
tensor = torch.tensor([1, 2, 3], device="cuda") #note it's not "gpu", but "cuda"
tensor, tensor.device

(tensor([1, 2, 3], device='cuda:0'), device(type='cuda', index=0))

In [12]:
tensor2 = torch.tensor([3, 2, 1])
tensor2.device

device(type='cpu')

In [14]:
tensor2.to(device)
tensor2.device

device(type='cpu')

In [15]:
tensor2 = tensor2.to(device)
tensor2

tensor([3, 2, 1], device='cuda:0')

In [16]:
tensor2_np = tensor2.numpy() # You can't do this without first passing it to CPU

TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

In [21]:
tensor2_np = tensor2.cpu().numpy()
tensor2_np.device, tensor2_np, tensor2.dtype

('cpu', array([3, 2, 1]), torch.int64)

### Exercises

1. Documentation reading - A big part of deep learning (and learning to code in general) is getting familiar with the documentation of a certain framework you're using. We'll be using the PyTorch documentation a lot throughout the rest of this course. So I'd recommend spending 10-minutes reading the following (it's okay if you don't get some things for now, the focus is not yet full understanding, it's awareness). See the documentation on torch.Tensor and for torch.cuda.
2. Create a random tensor with shape (7, 7).
3. Perform a matrix multiplication on the tensor from 2 with another random tensor with shape (1, 7) (hint: you may have to transpose the second tensor).
4. Set the random seed to 0 and do exercises 2 & 3 over again.
5. Speaking of random seeds, we saw how to set it with torch.manual_seed() but is there a GPU equivalent? (hint: you'll need to look into the documentation for torch.cuda for this one). If there is, set the GPU random seed to 1234.
6. Create two random tensors of shape (2, 3) and send them both to the GPU (you'll need access to a GPU for this). Set torch.manual_seed(1234) when creating the tensors (this doesn't have to be the GPU random seed).
7. Perform a matrix multiplication on the tensors you created in 6 (again, you may have to adjust the shapes of one of the tensors).
8. Find the maximum and minimum values of the output of 7.
9. Find the maximum and minimum index values of the output of 7.
10. Make a random tensor with shape (1, 1, 1, 10) and then create a new tensor with all the 1 dimensions removed to be left with a tensor of shape (10). Set the seed to 7 when you create it and print out the first tensor and it's shape as well as the second tensor and it's shape.

## Extra-curriculum
- Spend 1-hour going through the PyTorch basics tutorial (I'd recommend the Quickstart and Tensors sections).
- To learn more on how a tensor can represent data, see this video: What's a tensor?

## 1. torch.Tensor and torch.cuda

## torch.Tensor
Multi-dim matrix, containing elements of a single data type.

Data Types

Avoiding copies seems like another important aspect, probably due to the toll it takes to create a copy of big tensors.

Arguments: dtype and device

x.item() lets you retriee a Python number from a tensor containing a single value.

requires_grad True/False can be changed. Seems like it might be False by default.

## torch.Storage
Each tensor has one, "which holds its data". The tensor class provides a multi-dimensional, "strided" view of a storage, and defines numeric operations on it.

"In computer programming, the stride of an array (also referred to as increment, pitch or step size) is the number of locations in memory between beginnings of successive array elements, measured in bytes or in units of the size of the array's elements. The stride cannot be smaller than the element size but can be larger, indicating extra space between elements.

An array with stride of exactly the same size as the size of each of its elements is contiguous in memory. Such arrays are sometimes said to have unit stride. Unit stride arrays are sometimes more efficient than non-unit stride arrays, but non-unit stride arrays can be more efficient for 2D or multi-dimensional arrays, depending on the effects of caching and the access patterns used. This can be attributed to the principle of locality, specifically spatial locality."

According to the use cases Wikipedia lists, seems like it may be for "Array Cross-section", which requires a "Non-unit stride". You can see an example in which they access a column like this:

some_array(*, 2)

Very similar to what PyTorch lets us do.

## Memory overhead

It says here to prefer one large structure over many tiny tensors, since torch.Tensor, in its current implementation, introduces a memory overhead.

Needless to say, the documentation for torch.tensor is quite long, and has many, many details to talk about.

# 1. torch.cuda

Used to set up and run CUDA operations.
Keeps track fo the currently selected GPU, and all CUDA tensors you allocate will, by default, be created on that device.

Once a tensor is allocated, you can do operations on it irrespective of the selected device, and the result will always be placed on the same device as the tensor.

## Cross-GPU Operations

These are not allowed by default, with the exception of
- copy_()
- to()
- cuda()

These are copying or copy-like methods. Unless you enable peer-to-peer memory access, any attempts to launch ops on tensors spread across different devices will raise an error.

I see they operate on a specific device like so:

    with torch.cuda.device(1):
      a = torch.tensor([1., 2.], device=cuda)
      b = torch.tensor([1., 2.]).cuda()
      c = a + b --> Here c is on device 1
      z = x + y --> Here z is on device 0, as 'x' and 'y' are on device 0

      d = torch.randn(2, device=cuda2) #Note you can specify a device either way.


  

Seems like this article is more getting to the nuts and bolts of it, since it has a section on...

### TensorFloat-32
In essence to use TensorFloat32 (TF32) tensor cores, designed to achieve better performance on matmul and convolutions on torch.float32 tensors by rounding input data to have 10 bits of mantissa, and accumulating results with FP32 precision, maintaining FP32 dynamic range.





### 2. Random Tensor

In [3]:
import torch
tensor = torch.rand((7,7))
tensor.shape, tensor.ndim, tensor

(torch.Size([7, 7]),
 2,
 tensor([[0.2975, 0.7108, 0.1727, 0.0204, 0.8336, 0.7027, 0.2314],
         [0.4724, 0.0323, 0.4905, 0.6882, 0.4988, 0.8682, 0.6534],
         [0.2006, 0.0078, 0.6775, 0.6320, 0.4259, 0.0625, 0.5955],
         [0.4441, 0.0274, 0.6319, 0.5724, 0.1268, 0.3867, 0.2705],
         [0.6500, 0.4771, 0.3936, 0.6538, 0.8692, 0.2741, 0.3280],
         [0.1345, 0.7570, 0.1303, 0.0186, 0.9914, 0.1256, 0.5828],
         [0.8464, 0.2511, 0.6839, 0.5017, 0.9057, 0.2241, 0.3494]]))

### 3. Matmul of previous tensor with another random tensor with shape (1,7)

In [4]:
result = tensor @ torch.rand((1,7)).T #(7,7)*(7,1) = (7,1)
result.shape, result.ndim, result

(torch.Size([7, 1]),
 2,
 tensor([[1.7250],
         [2.0569],
         [1.4315],
         [1.1904],
         [1.7160],
         [1.6314],
         [1.9350]]))

### 4. Set the seed to 0 and do it again

In [8]:
torch.manual_seed(seed=0)
tensor = torch.rand((7,7))
tensor2 = torch.rand((1,7))
result = tensor @ tensor2.T
result

tensor([[1.8542],
        [1.9611],
        [2.2884],
        [3.0481],
        [1.7067],
        [2.5290],
        [1.7989]])

### 5. Set the Random seed on CUDA

In [1]:
!nvidia-smi

Wed Jun  4 02:22:26 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   65C    P8             13W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [3]:
import torch
torch.cuda.manual_seed(1234)

### 6. Random tensors of shape (2, 3) and send them both to the GPU

In [6]:
import torch
gpu_tensor = torch.rand((2,3), device="cuda")
gpu_tensor2 = torch.rand((2,3), device="cuda")
gpu_tensor.device, gpu_tensor2.device

(device(type='cuda', index=0), device(type='cuda', index=0))

### 7. Perform a Matrix Multiplication on the Tensors you created in 6

In [8]:
result = gpu_tensor @ gpu_tensor2.T
result

tensor([[0.5561, 0.2538],
        [1.2324, 0.9169]], device='cuda:0')

### 8. Find the minimum and maximum values of the output you just obtained

In [9]:
result.min(), result.max()

(tensor(0.2538, device='cuda:0'), tensor(1.2324, device='cuda:0'))

### 9. Obtain the max and min indices for the output in 7.

In [11]:
result.argmin().item(), result.argmax().item()

(1, 2)

### 10. Make a random tensor with shape (1, 1, 1, 10) and then create a new tensor with all the 1 dimensions removed to be left with a tensor of shape (10). Set the seed to 7 when you create it, and print out the first tensor, and its shape, as well as the second tensor, and its shape.

In [12]:
torch.cuda.manual_seed(7)
tensor = torch.rand((1, 1, 1, 10))
tensor2 = tensor.squeeze()
tensor, tensor.shape, tensor2, tensor2.shape

(tensor([[[[0.7574, 0.8106, 0.7558, 0.4035, 0.1435, 0.8589, 0.9435, 0.5754,
            0.7792, 0.7084]]]]),
 torch.Size([1, 1, 1, 10]),
 tensor([0.7574, 0.8106, 0.7558, 0.4035, 0.1435, 0.8589, 0.9435, 0.5754, 0.7792,
         0.7084]),
 torch.Size([10]))

### Extra-Curriculum

### What is a Tensor?

## Vectors
You may think of it as an arrow with magnitude and direction.
orientation - direction
magnitude - length of arrow

It may represent velocity, strength, ...
Vectors can represent an AREA as well, but how?

You simply make the length of the vector proportional to the area, and then you make the direction of the arrow to be PERPENDICULAR to the square.

Vectors are member of a wider class of object named TENSORS.

## Vector Components and Basis Vectors

Look at the three sticks that usually are taken to represent the X,Y,Z axes. It represents a COORDINATE SYSTEM. The Cartesian Coordinate System.

Coordinate systems come with basis vectors, or UNIT VECTORS. They have a length of 1

1 is whatever the unit you're expressing your vectors in.
x_hat or i_hat. It points in the direction of increasing X coordinates

y_hat
z_hat or k_hat

you get the idea. Coordinate System; Unit Vectors.

Now you can find the COMPONENTS of your vectors.

## Vector Components
Finding Vector Components.

Begin with a vector in the XY plane. To find the X component, project the vector into the X axis. To find the Y component, project the vector into the Y axis.

You 'shine a light' perpendicular to the X and Y axes, respectively. Notice that the inclination of the arrow changes the length of the shadow.

Ask yourself, to get from the base to the tip of the vector, how far do I have to go in the X vector, and how far in the Y vector.
How many x_hat units, how many y_hat units? (which are the basis vectors)

That means that instead of drawing an arrow for this vectors, you can say
4 x_hat, and 3 y_hat.

This is a perfectly valid representation of the vectors.
A = [[4], [3], [0]]

## Combinations of Components and Basis Vectors
All observers in all reference frames agree, not on basis vectors or components, but on their combination.

That is because basis vectors transform one way between reference frames.

AND the components transform in just such a way so as to keep the combination of these the same for all observers.

These characteristics led Lillian Lieber to call tensors the facts of the universe.


### Pytorch Basics tutorial

In [1]:
import torch
from torch import nn
from torch.utils.data import DataLoader #If you've played with fastai this seems familiar...
from torchvision import datasets
from torchvision.transforms import ToTensor

TorchText, TorchVision, TorchAudio, all include datasets. They are domain-specific libraries.

- torchvision.datasets
contains CIFAR, COCO, and others. Here FashionMNIST is used.

- Every TorchVision 'Dataset' includes two arguments:
  - transform: modifies the samples
  - target_transform: modifies the labels


In [2]:
# Downloading training data from open datasets
training_data = datasets.FashionMNIST(
    root='data',
    train=True,
    download=True,
    transform=ToTensor()
)

100%|██████████| 26.4M/26.4M [00:01<00:00, 14.3MB/s]
100%|██████████| 29.5k/29.5k [00:00<00:00, 210kB/s]
100%|██████████| 4.42M/4.42M [00:01<00:00, 3.89MB/s]
100%|██████████| 5.15k/5.15k [00:00<00:00, 20.4MB/s]


In [3]:
# Downloading test data from open datasets
test_data = datasets.FashionMNIST(
    root='data',
    train=False,
    download=True,
    transform=ToTensor()
)

We got our datasets, and now we must pass them as an argument to our DataLoader.

"This wraps an iterable over our dataset, and supports automatic batching, sampling, shuffling, and multiprocess data loading."

"Here we define a Batch Size of 64, i.e., each element in the DataLoader iterable will return a batch of 64 features and labels."

In [4]:
batch_size = 64

#Create data loaders
train_dataloader = DataLoader(
    dataset=training_data,
    batch_size=batch_size
)

test_dataloader = DataLoader(
    dataset=test_data,
    batch_size=batch_size
)

for X, y in test_dataloader:
  print(f"Shape of X [N, C, H, W]: {X.shape}")
  print(f"Shape of y: {y.shape}, {y.dtype}")
  break

Shape of X [N, C, H, W]: torch.Size([64, 1, 28, 28])
Shape of y: torch.Size([64]), torch.int64


### Creating a model
To define a Neural Network in PyTorch, we create a class that inherits from nn.Module

We define the layers of the network in the '__init__' function,
and specify how data will pass through the network in the 'forward' function.

To accelerate operations in the NN, we move it to the Accelerator, such as CUDA, MPS, MTIA, or XPU.

Not the device-agnostic code, which will default to CPU if no Accelerator is available.

In [5]:
device = torch.accelerator.current_accelerator().type if torch.accelerator.is_available() else "cpu"
print(f"Using {device} device")

Using cuda device


In [7]:
class NeuralNetwork(nn.Module):
  def __init__(self):
    super().__init__()
    self.flatten = nn.Flatten()
    self.linear_relu_stack = nn.Sequential( #This must be like the API for creating an acyclic graph
      nn.Linear(in_features=28*28, out_features=512), #Images are 28*28. And the output is a design choice.
      nn.ReLU(),
      nn.Linear(512, 512),
      nn.ReLU(),
      nn.Linear(512,10)
    )

  def forward(self, x):
    x = self.flatten(x)
    logits = self.linear_relu_stack(x)
    return logits

model = NeuralNetwork().to(device)
print(model)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (0): Linear(in_features=784, out_features=512, bias=True)
    (1): ReLU()
    (2): Linear(in_features=512, out_features=512, bias=True)
    (3): ReLU()
    (4): Linear(in_features=512, out_features=10, bias=True)
  )
)


### Optimizing Model Parameters
To train a model, we need:
- a Loss Function
- an Optimizer

In [11]:
loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(
    model.parameters(),
    lr=1e-3
)

In a Single Training Loop, the model makes predictions on the Training Dataset (fed to it in batches), and backpropagates the prediction error to adjust the model's parameters.

In [8]:
def train(dataloader, model, loss_fn, optimizer):
  size = len(dataloader.dataset)
  model.train()
  for batch, (X, y) in enumerate(dataloader):
    X, y = X.to(device), y.to(device)

    #Compute prediction error
    pred = model(X)
    loss = loss_fn(pred, y)

    # Backpropagation
    loss.backward() #Populates model's .grad fields, preparing for the optimization!
    optimizer.step()  # Uses the gradients (.grad field, calculated by the loss function) to update the parameters
    optimizer.zero_grad() # We reset for the next iteration (reset .grad)

    if batch % 100 == 0:
      loss, current = loss.item(), (batch + 1) * len(X) #Current loss, and number of samples processed so far (batch# * length of current batch, all are the same length)
      print(f"loss: {loss:>7f} [{current:>5d}/{size:>5d}]")

We must also check the model's performance against the Test Dataset to ensure it is learning

In [9]:
def test(dataloader, model, loss_fn):
  size = len(dataloader.dataset)
  num_batches = len(dataloader) #In PyTorch it's common to have two DataLoaders: one for the Training Dataset, one for the Test Dataset.
  model.eval() #Tells your model to change from TRAINING MODE to EVALUATION MODE. The behavior of some layers changes (Dropout, BatchNorm). We don't seem to have any of these, but it might have been included as an example!
  test_loss, correct = 0, 0
  with torch.no_grad(): #Disable automatic gradient tracking. PyTorch records the sequence of operations (like addition, multiplication, layers, etc.) as a graph. Then loss.backward() is a backward traversal of this graph.
    for X, y in dataloader:
      X, y = X.to(device), y.to(device)
      pred = model(X)
      test_loss += loss_fn(pred, y).item()
      correct += (pred.argmax(1) == y).type(torch.float).sum().item()

  test_loss /= num_batches # Loss per batch I think?
  correct /= size #Would give you a ratio of correct guesses
  print(f"Test Error: \n Accuracy: {(100*correct):>0.1f}%, Avg Loss: {test_loss:>8f} \n")

The training process is conducted over several iterations (epochs). During each epoch, the model learns PARAMETERS to make better predictions.

we print the model's accuracy and loss at each epoch; we'd like to see the acuracy INCREASE, and the loss DECREASE with every epoch

In [12]:
epochs = 5
for t in range(epochs):
  print(f"Epoch {t+1}\n-----------------------")
  train(
      dataloader=train_dataloader,
      model=model,
      loss_fn=loss_fn,
      optimizer=optimizer
  )
  test(
      test_dataloader,
      model,
      loss_fn
  )
print("Done!")

Epoch 1
-----------------------
loss: 2.308612 [   64/60000]
loss: 2.293439 [ 6464/60000]
loss: 2.273103 [12864/60000]
loss: 2.270180 [19264/60000]
loss: 2.248825 [25664/60000]
loss: 2.219795 [32064/60000]
loss: 2.229899 [38464/60000]
loss: 2.196112 [44864/60000]
loss: 2.199549 [51264/60000]
loss: 2.161415 [57664/60000]
Test Error: 
 Accuracy: 40.7%, Avg Loss: 2.155678 

Epoch 2
-----------------------
loss: 2.169899 [   64/60000]
loss: 2.154719 [ 6464/60000]
loss: 2.099880 [12864/60000]
loss: 2.115832 [19264/60000]
loss: 2.053022 [25664/60000]
loss: 1.998653 [32064/60000]
loss: 2.026659 [38464/60000]
loss: 1.948124 [44864/60000]
loss: 1.955741 [51264/60000]
loss: 1.877974 [57664/60000]
Test Error: 
 Accuracy: 53.1%, Avg Loss: 1.878342 

Epoch 3
-----------------------
loss: 1.916521 [   64/60000]
loss: 1.880146 [ 6464/60000]
loss: 1.768099 [12864/60000]
loss: 1.804783 [19264/60000]
loss: 1.686298 [25664/60000]
loss: 1.645986 [32064/60000]
loss: 1.663968 [38464/60000]
loss: 1.570072 [4

### Saving your Model

Serialize the internal state dictionary (which contains the model parameters)

In [13]:
torch.save(model.state_dict(), "model.pth")
print("Saved PyTorch model state to 'model.pth'")

Saved PyTorch model state to 'model.pth'


### Loading your Model

You need to re-create the model structure, then load the state dictionary into it.

In [14]:
model2 = NeuralNetwork().to(device)
model2.load_state_dict(
    state_dict=torch.load("model.pth", weights_only=True)
)

<All keys matched successfully>

Now you can use your model to make predictions

In [15]:
classes = [
    "T-shirt/top",
    "Trouser",
    "Pullover",
    "Dress",
    "Coat",
    "Sandal",
    "Shirt",
    "Sneaker",
    "Bag",
    "Ankle boot",
]

model2.eval()
x, y = test_data[0][0], test_data[0][1]
with torch.no_grad():
  x = x.to(device)
  pred = model(x)
  predicted, actual = classes[pred[0].argmax(0)], classes[y]
  print(f"Predicted: {predicted}, Actual: {actual}")

Predicted: Ankle boot, Actual: Ankle boot


### Tensors (Second tutorial)

Encodes inputs and outputs of a model, but model's parameters as well.

Similar to numpy's ndarrays, except that tensors can run on GPUs or other hardware accelerators.

Tensors and NumPy arrays can often share the same underlying memory, eliminating the need to copy data.

Tensors are also optimized for Automatic Differentiation (Autograd).

If you're familiar with ndarrays, you'll be right at home with the Tensor API.